---
title: "Analyzing_RNAseq_data_DESeq2"
output: 
  pdf_document: 
    latex_engine: xelatex
    keep_tex: yes
---



# Analyzing RNA-seq data with DESeq2



## Loading library

```{r}
library("DESeq2")
library("apeglm")
library("ggplot2")
library("vsn")
library("pheatmap")
library("RColorBrewer")
```

## Preparation 

### Set the Directory

The folder has been uploaded to Google Drive

```{r}
#directory <- "/Users/margery/Desktop/data/New_Lobular_Duct/" 


# In google Drivw
direction<- "~/510_Final_Project_Data/New_Lobular_Duct"
```


### Set the `sample condition` & `sampleFiles`

Extract the information of Condition directly from file names which could guarantee the one-to-one correspondence of expressed matrix and samples

Use `grep` to select those files containing string `group`
Use `sub` to chop up the sample filename to obtain the condition status
```{r}
sampleFiles <- grep("group",list.files(directory),value=TRUE)
sampleCondition <- sub("(.*group).*","\\1",sampleFiles)
```

### Set `sampleTable`
Only consider the `condition` as Factor

```{r}
sampleTable <- data.frame(sampleName = sampleFiles,
                          fileName = sampleFiles,
                          condition = sampleCondition)
sampleTable$condition <- factor(sampleTable$condition)

```






# Build the DESeqDataSet （ Only Factor is "condition" )

```{r}
dds <- DESeqDataSetFromHTSeqCount(sampleTable = sampleTable,
                                       directory = directory,
                                       design= ~ condition)
dds

```
### Remove the suffix of fileName in `sampleTable`

```{r}
library(tidyr)
sampleTable <- sampleTable %>%
  tidyr::separate(fileName,into = c("fileName"),sep = "\\.")
```



By default, R will choose a reference level for factors based on alphabetical order

```{r}
head(dds$condition)
```
In this case , Ductgroup is the reference , define it manually to make sure

log2 fold change and Wald test p value ： last level / reference level
log2 fold change ： log2 (Lobulargroupt / Ductgroup)


```{r}
Col_Duct_Lobular <- read.csv("~/Files/Col_Duct_Lobular.csv")
```


## Build sampleTable with 2 Factors

Both `condition` & `Stage` are Factors



```{r}
colnames(Col_Duct_Lobular)[2] <- "fileName"
sampleTable <- merge(sampleTable,Col_Duct_Lobular,by="fileName")
sampleTableselect <- sampleTable
```

###Add multiple factors "Stage"

Select Necessary Columns

```{r}
library(dplyr)
sampleTableselect <- sampleTableselect%>%
  dplyr::select(fileName,sampleName,condition,ajcc_pathologic_stage)
```

```{r}
sampleTableselect$condition <- factor(sampleTableselect$condition)
sampleTableselect$Stage <- factor(sampleTableselect$ajcc_pathologic_stage)
```

# Build the DESeqDataSet （ Factors are "condition" and "Stage" )

```{r}
ddsMF <- DESeqDataSetFromHTSeqCount(sampleTable = sampleTableselect,
                                       directory = directory,
                                       design= ~ condition + Stage)
ddsMF

```

```{r}
ddsMF <- DESeq(ddsMF)
head(ddsMF)
```


## Pre-filtering

Remove the genes with few reads (less than 10 ) to reduce the memory size and increase the speed 
```{r}
keep <- rowSums(counts(dds)) >= 10
dds <- dds[keep,]
dds
```

```{r}
dds
```

After filtering, the number of elements has decreased from 60483 to 50442

### Check Factors

```{r}
head(ddsMF$condition)
head(ddsMF$Stage)
```


# Differential expression analysis

Use function `DESeq` to do differential expression analysis
Use function `results` to generate results tables with log2 fold changes, p values and adjusted p values

```{r}
dds <- DESeq(dds)
res <- results(dds)
```

```{r}
res
```

***Export the results to csv file***

```{r}
write.csv(res,file="~/Results/Res_Lobular_Duct_All.csv")
```


## Log fold change shrinkage for visualization and ranking

Use function `lfcShrink` to shrink the LFC
`apeglm`: (Zhu, Ibrahim, and Love 2018) effect size shrinkage,which improves on the previous estimator

```{r}
resultsNames(dds)
```

```{r}
resLFC <- lfcShrink(dds, coef="condition_Lobulargroup_vs_Ductgroup", type="apeglm")
resLFC
```

`resLFC` is more compacted compared to `res`
column `stat` is removed after shrinking 

```{r}
names(resLFC)
names(res)
```

## Speed up 

Use `parallel=TRUE` and `BPPARAM=MulticoreParam(4)` to split the job over 4 cores

```{r}
library("BiocParallel")
register(MulticoreParam(4))
```


## Define the Differential Expressed Gene and Export the results

Define DEG as padj <= 0.05 & abs(log2FoldChange) >= 1.5

```{r}
diff_gene_deseq2 <-subset(res, padj <= 0.05 & abs(log2FoldChange) >= 1.5)
dim(diff_gene_deseq2)
head(diff_gene_deseq2)
```


```{r}
summary(diff_gene_deseq2)
```

```{r}
write.csv(diff_gene_deseq2,file = "New_DEG_Lobular_Duct.csv")
```



```{r}
diff_gene_deseq2 <- read.csv("~/Results/New_DEG_Lobular_Duct.csv")
```

## ID Transfer
### Change column name
```{r}
colnames(diff_gene_deseq2)[1] <- "gene_id"
```



### Remove the version number
```{r}
library(tidyr)
diff_gene_deseq2 <- diff_gene_deseq2 %>%
  tidyr::separate(gene_id,into = c("gene_id"),sep = "\\.")
```

```{r}
library(AnnotationDbi)
library(org.Hs.eg.db)
diff_gene_deseq2$symbol <- mapIds(org.Hs.eg.db,
                     keys=diff_gene_deseq2$gene_id,
                     column="SYMBOL",
                     keytype="ENSEMBL",
                     multiVals="first")
```
### Remove Duplicated and NA

```{r}
library(dplyr)
diff_gene_deseq2 <- diff_gene_deseq2 %>% 
  ## Remove NA
  filter(symbol!="NA") %>% 
  ## Remove Duplicate
  distinct(symbol,.keep_all = T)
```

### Extract UP-regulated & Down-regulated

```{r}
Up_diff <- data.frame(subset(diff_gene_deseq2,log2FoldChange > 0) )
Up_diff$gene_reg <- "Up"
Down_diff <- data.frame(subset(diff_gene_deseq2,log2FoldChange < 0) )
Down_diff$gene_reg <- "Down"
  
```
```{r}
reg_diff <- rbind(Up_diff,Down_diff)
```


```{r}
write.csv(Up_diff,"～/Results/Up_diff.csv",row.names = F)
write.csv(Down_diff,"～/Results/Down_diff.csv",row.names = F)
write.csv(reg_diff,"～/Results/All_diff_Reg.csv",row.names = F)
```


## More information on results columns 

Use function `mcols` to find which variables and tests were used 

```{r}
mcols(res)$description
```

For a particular gene, a log2 fold change of -1 for `condition logroup vs ductgroup` means that the `logroup` induces a multiplicative change in observed gene expression level of 2^-1 = 0.5 compared to the untreated condition. 

# Visulization 

## MA-plot 

`plotMA` shows the log2 fold changes attributable to a given variable over the mean of normalized counts for all the samples in the DESeqDataSet.
```{r}
plotMA(res, ylim=c(-2,2))
```



```{r}
plotMA(resLFC, ylim=c(-2,2))
```


```{r}
library(ashr)
resAsh <- lfcShrink(dds, coef=2, type="ashr")
# resNorm <- lfcShrink(dds, coef=2, type="normal") TOO Slow 

```

```{r}
par(mfrow=c(1,3), mar=c(4,4,2,1))
xlim <- c(1,1e5); ylim <- c(-3,3)
plotMA(resLFC, xlim=xlim, ylim=ylim, main="apeglm")
#plotMA(resNorm, xlim=xlim, ylim=ylim, main="normal") too Slow
plotMA(resAsh, xlim=xlim, ylim=ylim, main="ashr")
```
## Plot Counts

Examine the counts of reads for a single gene across the groups 

Use function `plotCounts` to normalize counts by the estimated size factors and adds a pseudocount of 1/2 to allow for log scale plotting.



Here I sepcify the gene `MAGEA4` which is found Down-Regulataed 

In IPA analysis, this gene is in the Pathway of disease "HER2 non-overexpressing breast carcinoma"

For customized plotting, an argument `returnData` specifies that the function should only return a data.frame for plotting with ggplot.

MAGEA4

```{r}
d <- plotCounts(dds, gene="ENSG00000147381.10", intgroup="condition", 
                returnData=TRUE)
library("ggplot2")
ggplot(d, aes(x=condition, y=count)) + ggtitle("MAGEA4")+
  geom_point(position=position_jitter(w=0.1,h=0)) + 
  scale_y_log10(breaks=c(25,100,400))
```





## BOX plot


```{r}


d1 <- plotCounts(dds,gene="ENSG00000147381.10", intgroup="condition",returnData = T)
              
ggplot(d1,aes(condition, count)) + geom_boxplot(aes(fill=condition)) + scale_y_log10()

```

## Count data Transformation


```{r}
vsd <- vst(dds, blind=FALSE)
```

```{r}
# this gives log2(n + 1)
ntd <- normTransform(dds)
```

```{r}
library("vsn")
meanSdPlot(assay(ntd))

```
```{r}
meanSdPlot(assay(vsd))
```


# Data transformations and visualization (Use ddsMF)

## Count data transformations


Use function `vsd` to remove the dependence of the variance on the mean instead of function `rlog` for it takes MUCH less time        

Usually use `vsd` if the number of samples > 50

```{r}
vsdMF <- vst(ddsMF, blind=FALSE)
```

```{r}
ntdMF <- normTransform(ddsMF)
```

```{r}
df <- sampleTableselect %>%
  dplyr::select(fileName,condition,Stage)
```

```{r}
rownames(df)<-df[,1]
 df <- df[,-1]
```


## Heatmap

```{r}
library("pheatmap")
select <- order(rowMeans(counts(ddsMF,normalized=TRUE)),
                decreasing=TRUE)[1:20]
```
```{r}
pheatmap(assay(ntdMF)[select,], cluster_rows=FALSE,show_colnames=FALSE,
         cluster_cols=FALSE,annotation = df,fontsize_row = 6)
```

```{r}
pheatmap(assay(vsdMF)[select,], cluster_rows=FALSE,show_colnames=FALSE,
         cluster_cols=FALSE,annotation = df,fontsize_row = 6)
```

## Heatmap of the sample-to-sample distances

Apply the dist function to the transpose of the transformed count matrix to get sample-to-sample distances.
```{r}
sampleDists <- dist(t(assay(vsdMF)))

```

```{r}
library("RColorBrewer")
sampleDistMatrix <- as.matrix(sampleDists)
rownames(sampleDistMatrix) <- paste(vsdMF$condition, vsdMF$type,vsdMF$Stage, sep="-")
colnames(sampleDistMatrix) <- NULL
colors <- colorRampPalette( rev(brewer.pal(9, "Blues")) )(255)

```
```{r}
pheatmap(sampleDistMatrix,
         clustering_distance_rows=sampleDists,
         clustering_distance_cols=sampleDists,
         col=colors,show_rownames = F)
```

## PCA plot

```{r}
plotPCA(vsdMF, intgroup=c("condition","Stage"))
```

```{r}
library(pcaExplorer)
pcaplot(vsdMF,intgroup = c("condition","Stage"),ntop = 1000,
        pcX = 1, pcY = 2, title = "pcaplot-vsdMF",
        ellipse = TRUE)
```


```{r}
plotPCA(ntdMF, intgroup=c("condition","Stage"))
```


```{r}
pcaplot(ntdMF,intgroup = c("condition","Stage"),ntop = 1000,
        pcX = 1, pcY = 2, title = "pcaplot-ntdMF",
        ellipse = TRUE)
```

