---
title: "R Notebook"
output:
  html_document:
    df_print: paged
  html_notebook: default
  pdf_document: default
---

## Loading library

```{r}
library("DESeq2")
library("apeglm")
library("ggplot2")
library("vsn")
library("pheatmap")
library("RColorBrewer")
```





 **The names of files have been changed by bash**: 

add suffix "Ductgroup" in the htseqcounts files of Duct group

add suffix "Lobulargroup" in the htseqcounts files of Lobular group

put them together in one folder named `Lobular_Ductr` to point it to this directory 



## Preparation 

### Set the Directory

The folder has been uploaded to Google Drive

```{r}
directory <- "/Users/margery/Desktop/data/Lobular_Duct/" 
#in my computer

#direction<- "~/TRGN510_Final_Project_Data/Lobular_Duct"
```

### Set the `sample condition` & `sampleFiles`

Extract the information of Condition directly from file names which could guarantee the one-to-one correspondence of expressed matrix and samples

Use `grep` to select those files containing string `group`
Use `sub` to chop up the sample filename to obtain the condition status
```{r}
sampleFiles <- grep("group",list.files(directory),value=TRUE)
```
```{r}
sampleCondition <- sub("(.*group).*","\\1",sampleFiles)
```





# Build the DESeqDataSet

```{r}
sampleTable <- data.frame(sampleName = sampleFiles,
                          fileName = sampleFiles,
                          condition = sampleCondition)
sampleTable$condition <- factor(sampleTable$condition)
```

```{r}
library("DESeq2")
dds <- DESeqDataSetFromHTSeqCount(sampleTable = sampleTable,
                                       directory = directory,
                                       design= ~ condition)
dds

```
60483

## Pre-filtering

Remove the genes with few reads (less than 10 ) to reduce the memory size and increase the speed 
```{r}
keep <- rowSums(counts(dds)) >= 10
dds <- dds[keep,]

```

After filtering, the number of elements has decreased from 60483 to 50860

By default, R will choose a reference level for factors based on alphabetical order

```{r}
dds$condition
```
In this case , Ductgroup is the reference , define it manually to make sure

log2 fold change and Wald test p value ： last level / reference level
log2 fold change ： log2 (Lobulargroupt / Ductgroup)

# Differential expression analysis

Use function `DESeq` to do differential expression analysis
Use function `results` to generate results tables with log2 fold changes, p values and adjusted p values.

```{r}
dds <- DESeq(dds)
res <- results(dds)
```

```{r}
res
```

***Export the results to csv file***
```{r}
write.csv(res,file="Lobular_Duct_res.csv")
```


## Log fold change shrinkage for visualization and ranking

Use function `lfcShrink` to shrink the LFC
`apeglm`: (Zhu, Ibrahim, and Love 2018) effect size shrinkage,which improves on the previous estimator

```{r}
resultsNames(dds)
```

```{r}
resLFC <- lfcShrink(dds, coef="condition_Lobulargroup_vs_Ductgroup", type="apeglm")
resLFC
```

`resLFC` is more compacted compared to `res`
column `stat` is removed after shrinking 

```{r}
names(resLFC)
names(res)
```

## Speed up 

Use `parallel=TRUE` and `BPPARAM=MulticoreParam(4)` to split the job over 4 cores

```{r}
library("BiocParallel")
register(MulticoreParam(4))
```

## p-values and adjusted p-values

order our results table by the smallest p value 

```{r}
resOrdered <- res[order(res$pvalue),]
summary(res)
```

***Find out How many adjusted p-values were less than 0.05?***
```{r}
sum(res$padj < 0.05, na.rm =T)
```

By default the argument alpha is set to  0.1, set the alpha = 0.05 means set the statistical significance to be 0.05

```{r}
res05 <- results(dds, alpha=0.05)
summary(res05)
```

## Define the Differential Expressed Gene and Export the results

```{r}
diff_gene_deseq2 <-subset(res, padj < 0.05 & abs(log2FoldChange) > 1)
dim(diff_gene_deseq2)
head(diff_gene_deseq2)

```

```{r}
write.csv(diff_gene_deseq2,file = "DEG_Lobular_Duct.csv")
```

# Visulization 

## MA-plot 

`plotMA` shows the log2 fold changes attributable to a given variable over the mean of normalized counts for all the samples in the DESeqDataSet.
```{r}
plotMA(res, ylim=c(-2,2))
```



```{r}
plotMA(resLFC, ylim=c(-2,2))
```


```{r}
library(ashr)
resNorm  <- lfcShrink(dds, coef=2, type="normal")
resAsh <- lfcShrink(dds, coef=2, type="ashr")
```

```{r}
par(mfrow=c(1,3), mar=c(4,4,2,1))
xlim <- c(1,1e5); ylim <- c(-3,3)
plotMA(resLFC, xlim=xlim, ylim=ylim, main="apeglm")
plotMA(resNorm, xlim=xlim, ylim=ylim, main="normal")
plotMA(resAsh, xlim=xlim, ylim=ylim, main="ashr")
```






## Plot counts

Examine the counts of reads for a single gene across the groups 

Use function `plotCounts` to normalize counts by the estimated size factors and adds a pseudocount of 1/2 to allow for log scale plotting.

Here we specify the gene which had the **smallest p value** from the results table created above. 
```{r}
plotCounts(dds, gene=which.min(res$padj), intgroup="condition")
```

For customized plotting, an argument returnData specifies that the function should only return a data.frame for plotting with ggplot.
```{r}
d <- plotCounts(dds, gene=which.min(res$padj), intgroup="condition", 
                returnData=TRUE)
library("ggplot2")
ggplot(d, aes(x=condition, y=count)) + 
  geom_point(position=position_jitter(w=0.1,h=0)) + 
  scale_y_log10(breaks=c(25,100,400))
```

## More information on results columns 

Use function `mcols` to find which variables and tests were used 

```{r}
mcols(res)$description
```

For a particular gene, a log2 fold change of -1 for `condition logroup vs ductgroup` means that the `logroup` induces a multiplicative change in observed gene expression level of 2^-1 = 0.5 compared to the untreated condition. 

##
# Data transformations and visualization

## Count data transformations


Use function `vsd` to remove the dependence of the variance on the mean instead of function `rlog` for it takes MUCH less time        
```{r}
vsd <- vst(dds, blind=FALSE)
```

It takes more than 24hrs, so I cut it off
```{r}
#rld <- rlog(dds, blind=FALSE)
```


## Extracting transformed values

```{r}
# this gives log2(n + 1)
ntd <- normTransform(dds)

```
```{r}
library("vsn")
meanSdPlot(assay(ntd))

```
```{r}
meanSdPlot(assay(vsd))
```
# Data quality assessment by sample clustering and visualization

## Heatmap of the count matrix

```{r}

df <- as.data.frame(sampleCondition)

```


```{r}
library("pheatmap")
select <- order(rowMeans(counts(dds,normalized=TRUE)),
                decreasing=TRUE)[1:20]
```
```{r}
rownames(df) <- colnames(ntd)

```

```{r}
pheatmap(assay(ntd)[select,], cluster_rows=FALSE,show_colnames=FALSE,
         cluster_cols=FALSE,annotation = df,cellwidth=2)
```

```{r}
pheatmap(assay(vsd)[select,], cluster_rows=FALSE,show_colnames=FALSE,
         cluster_cols=FALSE,annotation = df,fontsize_row = 6)
```

```{r}
normCounts <- counts(dds,normalized = T)
```


## Heatmap of the sample-to-sample distances

Apply the dist function to the transpose of the transformed count matrix to get sample-to-sample distances.
```{r}
sampleDists <- dist(t(assay(vsd)))

```

```{r}
library("RColorBrewer")
sampleDistMatrix <- as.matrix(sampleDists)
rownames(sampleDistMatrix) <- paste(vsd$condition, vsd$type, sep="-")
colnames(sampleDistMatrix) <- NULL
colors <- colorRampPalette( rev(brewer.pal(9, "Blues")) )(255)

```
```{r}
pheatmap(sampleDistMatrix,
         clustering_distance_rows=sampleDists,
         clustering_distance_cols=sampleDists,
         col=colors,show_rownames = F)
```

## Principal component plot of the samples
```{r}
plotPCA(vsd, intgroup=c("condition"))
```
customize the PCA plot using the ggplot function.

```{r}
pcaData <- plotPCA(vsd, intgroup=c("condition"), returnData=TRUE)
percentVar <- round(100 * attr(pcaData, "percentVar"))
ggplot(pcaData, aes(PC1, PC2, color=condition, shape=condition)) +
  geom_point(size=2) +
  xlab(paste0("PC1: ",percentVar[1],"% variance")) +
  ylab(paste0("PC2: ",percentVar[2],"% variance")) + 
  coord_fixed()
```


# Filter DEG

```{r}
diff_gene_deseq2 <-subset(res, padj < 0.05 & abs(log2FoldChange) > 1)
dim(diff_gene_deseq2)
head(diff_gene_deseq2)

```
```{r}
write.csv(diff_gene_deseq2,file = "DEG_Lobular_Duct.csv")
```

