

# Map the Ensembl ID to gene ID in Genome


## Make readble gtf file in R
```{r}
if(!require("rtracklayer")) BiocManager::install("rtracklayer") 
gtf1 <- rtracklayer::import("Homo_sapiens.GRCh38.104.chr.gtf")
gtf_df <- as.data.frame(gtf1)
```


```{r}
save(gtf_df,file = "gtf_df.Rdata")
```

## Use AnnotationDbi package to do get Mapping 

```{r}
get_map = function(input) {
  if (is.character(input)) {
    if(!file.exists(input)) stop("Bad input file.")
    message("Treat input as file")
    input = data.table::fread(input, header = FALSE)
  } else {
    data.table::setDT(input)
  }
  
  input = input[input[[3]] == "gene", ]
  
  pattern_id = ".*gene_id \"([^;]+)\";.*"
  pattern_name = ".*gene_name \"([^;]+)\";.*"
  
  
  gene_id = sub(pattern_id, "\\1", input[[9]])
  gene_name = sub(pattern_name, "\\1", input[[9]])
  
  Ensembl_ID_TO_Genename <- data.frame(gene_id = gene_id,
                                        gene_name = gene_name,
                                        stringsAsFactors = FALSE)
  return(Ensembl_ID_TO_Genename)
}
```
```{r}
Ensembl_ID_TO_Genename <- get_map("gencode.v29.annotation.gtf") 
```


## Remove the version number
```{r}
gtf_Ensembl_ID <- substr(Ensembl_ID_TO_Genename[,1],1,15)
Ensembl_ID_TO_Genename <- data.frame(gtf_Ensembl_ID,Ensembl_ID_TO_Genename[,2])
```
```{r}
colnames(Ensembl_ID_TO_Genename) <- c("Ensembl_ID","gene_id")
```
```{r}
write.csv(Ensembl_ID_TO_Genename,file = "Ensembl_ID_TO_Genename.csv",row.names = F)
```



# Change the Ensemble ID to gene id in DEG

## Read DEG file 

```{r}
diff_gene_deseq2 <- read.csv("~/Results/New_DEG_Lobular_Duct.csv")
```

## Change column name
```{r}
colnames(diff_gene_deseq2)[1] <- "gene_id"
```



## Remove the version number
```{r}
library(tidyr)
diff_gene_deseq2 <- diff_gene_deseq2 %>%
  tidyr::separate(gene_id,into = c("gene_id"),sep = "\\.")
```

## Get gene symbol ID in DEG

```{r}
library(AnnotationDbi)
library(org.Hs.eg.db)
diff_gene_deseq2$symbol <- mapIds(org.Hs.eg.db,
                     keys=diff_gene_deseq2$gene_id,
                     column="SYMBOL",
                     keytype="ENSEMBL",
                     multiVals="first")
```


## Remove duplicated genes 

```{r}
library(dplyr)
diff_gene_deseq2 <- diff_gene_deseq2 %>% 
  ## Remove NA
  filter(symbol!="NA") %>% 
  ## Remove Duplicate
  distinct(symbol,.keep_all = T)
```

Number has been reduced to 309 from 463

## Save the correspondence between Gene id and Ensembl ID 

```{r}
DEG_Ensemble_Symbol <- diff_gene_deseq2[,-c(2:7)]
write_csv(DEG_Ensemble_Symbol, "DEG_Ensemble_Symbol.csv")
```


## Remove the ensemble ID 

```{r}
diff_gene_deseq2$gene_id <- diff_gene_deseq2$symbol
diff_gene_deseq2 <- diff_gene_deseq2[,-8]
```

## Export Results to csv file

```{r}
write.csv(diff_gene_deseq2,"New_symbolID_refiltered.csv",row.names = F)
```

